daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1316,6 +1316,173 @@ object SBuiltin {
+            // TODO [FM]: replace with ASCII-specific function, or not
+        case x =>
+          throw SErrorCrash(s"type mismatch SBTextToLower, expected Text got $x")
+      }
+    }
+  }
+
+  /** $text_slice :: Int -> Int -> Text -> Text */
+  final case object SBTextSlice extends SBuiltin(3) {
+    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+      args.get(0) match {
+        case SInt64(from) =>
+          args.get(1) match {
+            case SInt64(to) =>
+              args.get(2) match {
+                case SText(t) =>
+                  if (to <= 0 || from >= t.length.toLong || to <= from ) {
+                    machine.ctrl = CtrlValue(SText(""))
+                  } else {
+                    val rfrom = from.max(0).toInt
+                    val rto = to.min(t.length.toLong).toInt
+                      // NOTE [FM]: We use toInt only after ensuring the indices are
+                      // between 0 and t.length inclusive. Calling toInt prematurely
+                      // would mean dropping the high order bits indiscriminitely,
+                      // so for instance (0x100000000L).toInt == 0, resulting in an
+                      // empty string below even though `to` was larger than the
+                      // length.
+                    machine.ctrl = CtrlValue(SText(t.slice(rfrom, rto)))
+                  }
+                case x =>
+                  throw SErrorCrash(s"type mismatch SBTextSlice, expected Text got $x")
+              }
+            case x =>
+              throw SErrorCrash(s"type mismatch SBTextSlice, expected Int64 got $x")
+          }
+        case x =>
+          throw SErrorCrash(s"type mismatch SBTextSlice, expected Int64 got $x")
+      }
+    }
+  }
+
+  /** $text_slice_index :: Text -> Text -> Optional Int */
+  final case object SBTextSliceIndex extends SBuiltin(2) {
+    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+      args.get(0) match {
+        case SText(slice) =>
+          args.get(1) match {
+            case SText(t) =>
+              val n = t.indexOfSlice(slice)
+              if (n < 0) {
+                machine.ctrl = CtrlValue(SOptional(None))
+              } else {
+                machine.ctrl = CtrlValue(SOptional(Some(SInt64(n.toLong))))
+              }
+            case x =>
+              throw SErrorCrash(s"type mismatch SBTextSliceIndex, expected Text got $x")
+          }
+        case x =>
+          throw SErrorCrash(s"type mismatch SBTextSliceIndex, expected Text got $x")
+      }
+    }
+  }
+
+  /** $text_contains_only :: Text -> Text -> Bool */
+  final case object SBTextContainsOnly extends SBuiltin(2) {
+    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+      args.get(0) match {
+        case SText(alphabet) =>
+          args.get(1) match {
+            case SText(t) =>
+              val alphabetSet = alphabet.toSet
```

Comment:
```suggestion
              val alphabetSet = alphabet.codePoints().iterator().asScala.toSet
```

assuming you have the following line somewhere in the scope:
```
import scala.collection.JavaConverters._
```